tsdb/engine/tsm1/cache.go

```diff
@@ -81,6 +81,7 @@ const (
 type Cache struct {
 	mu      sync.RWMutex
 	store   map[string]*entry
+	dirty   map[string]*entry
```

Comment:
This analysis is not correct and is not consistent with the go memory model.

Cache.Deduplicate() executes without taking any kind of lock. To do what it needs to do it sometimes has to updte the e.values reference of some entries. This is done here:

https://github.com/jonseymour/influxdb/blob/f44db5df48b7eaa00ccbbc9b76406b29cf5a2bb3/tsdb/engine/tsm1/cache.go#L61-L62

While it is doing this, code that is running with Cache.merged() under either read or write lock is accessing the entries at line 302 of cache.go:

https://github.com/jonseymour/influxdb/blob/f44db5df48b7eaa00ccbbc9b76406b29cf5a2bb3/tsdb/engine/tsm1/cache.go#L302

So, in one instance we have a goroutine (c.merged) which is accessing the c.snapshot.store[k].value while properly protected by a lock (Cache.merged()) and another goroutine (Cache.Deduplicate) which is accessing the same entry c.snapshot.store[k] without any lock.

This exactly the condition that the go memory model requires the use of synchronization.

> When multiple goroutines access a shared variable v, they must use synchronization events to establish happens-before conditions that ensure reads observe the desired writes.

(https://golang.org/ref/mem)

This is also why the data race detector is reporting the existence of a data race.

Could I ask you to read the go memory model again, ask a colleague to read the same text and this PR and if you are still not convinced ask the #general channel gophers.slack.com for a second opinion. If I am wrong with my analysis then I am quite sure that everyone on that channel will be able to explain exactly why I am wrong.

 It was my understanding, for example, that @e-dard did accept the reality of the race condition which is demonstrated by the test in the previous commit.